units/src/fee.rs

```diff
@@ -146,28 +148,24 @@ impl FeeRate {
     /// fee is a non-integer amount, the amount is rounded up, ensuring that the transaction fee is
     /// enough instead of falling short if rounded down.
     ///
-    /// Returns [`None`] if overflow occurred.
-    #[must_use]
-    pub const fn checked_mul_by_weight(self, weight: Weight) -> Option<Amount> {
-        // No `?` operator in const context.
-        match self.to_sat_per_kwu().checked_mul(weight.to_wu()) {
-            Some(mul_res) => match mul_res.checked_add(999) {
-                Some(add_res) => match Amount::from_sat(add_res / 1000) {
-                    Ok(fee) => Some(fee),
-                    Err(_) => None,
-                },
-                None => None,
-            },
-            None => None,
+    /// Returns [`NumOpResult::Error`] if overflow occurred.
+    pub const fn checked_mul_by_weight(self, weight: Weight) -> NumOpResult<Amount> {
+        if let Some(fee) =  self.to_sat_per_kwu().checked_mul(weight.to_wu()) {
+            if let Some(round_up) = fee.checked_add(999) {
+                if let Ok(ret) = Amount::from_sat(round_up / 1_000) {
+                    return NumOpResult::Valid(ret);
+                }
+            }
```

Comment:
Oh, interesting. It appears as if this was lossy for some values but because we have the 21 M limit it's not.

A simpler version of this function would be:

```rust
let product = u128::from(self.to_sat_per_kwu()) * u128::from(weigth.to_wu());
if product > Amount::MAX.to_sat() * 1000 {
    Err(NumOpResult::Error(E::while_doing(MathOp::Mul)))
} else {
    // I wonder if the compiler can optimize the check here. I believe it should.
    Ok(Amount::from_sat(((product + 999) / 1000) as u64).expect("overfloww checked above"))
}
```

Reply:
I have some questions about this, and it's distinctly different than what was here before, so I think maybe it could be a followup commit.  I've not seen `u128` used in the code base which would allow bypassing a possible overflow check I guess by using more bits.  This data type must perform worse than a u64 though since most architectures are 64bit still?

nit you misspelled overflow :P

Reply:
It performs worse than `u64` on platforms that don't support it natively (many actually do - it has nothing to do with the pointer size) but I highly doubt it's worse than having *three* branches instead of one (at least it's more predictable perf).

Feel free to benchmark it if you want but unless the difference is large, I'd prefer cleaner code.

Reply:
Oh this code is the benchmark in question.  I think it would be hard to have a bechmark that's sensitive enough to show a difference although I'll take a closer look.

> It performs worse than u64 on platforms that don't support it natively (many actually do - it has nothing to do with the pointer size)

Are you referring to a PAE kernel?  I thought there was some kind of performance hit for this but maybe not.

Reply:
see https://github.com/rust-bitcoin/rust-bitcoin/pull/4480